### PR TITLE
Unify shared infra

### PR DESCRIPTION
**Pull request by alliepiper**
## Description

This PR centralizes shared infrastructure to the CCCL level, reducing redundancy across our projects and facilitating the introduction of new subprojects.

### CODE REVIEW

File: cmake/CCCLBuildCompilerTargets.cmake [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+#   be defined for each supported dialect.
+#
+# cccl.silence_unreachable_code_warnings
+# - Interface target that silences unreachable code warnings.
+# - Used to selectively disable such warnings in unit tests caused by
+#   unconditionally thrown exceptions.
+
+set(CCCL_KNOWN_CXX_DIALECTS 11 14 17 20)
+
+# sccache cannot handle the -Fd option generationg pdb files
```
Comment (miscco) [inline cmake/CCCLBuildCompilerTargets.cmake:20]:
```suggestion
# sccache cannot handle the -Fd option generating pdb files
```

File: cmake/CCCLBuildCompilerTargets.cmake [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    # See https://github.com/microsoft/STL/issues/696
+    append_option_if_available("/wd4494" cxx_compile_options)
+
+  else()
+    list(APPEND cuda_compile_options "-Wreorder")
+
+    append_option_if_available("-Werror" cxx_compile_options)
+    append_option_if_available("-Wall" cxx_compile_options)
+    append_option_if_available("-Wextra" cxx_compile_options)
+    append_option_if_available("-Wreorder" cxx_compile_options)
```
Comment (miscco) [inline cmake/CCCLBuildCompilerTargets.cmake:106]: Should we enable that one and see what happens?

File: cmake/CCCLBuildCompilerTargets.cmake [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    append_option_if_available("-Winit-self" cxx_compile_options)
```
Comment (miscco) [inline cmake/CCCLBuildCompilerTargets.cmake:107]: That also looks like a useful warning, can we enable it test wise?

File: cudax/cmake/cudaxBuildCompilerTargets.cmake [hunk 4]
```diff
@@ -1,113 +1,51 @@
...
-    # `__cudaUnregisterBinaryUtil`.
-    append_option_if_available("/wd4505" cxx_compile_options)
-    # C4706: assignment within conditional expression
-    # MSVC doesn't provide an opt-out for this warning when the assignment is
-    # intentional. Clang will warn for these, but suppresses the warning when
-    # double-parentheses are used around the assignment. We'll let Clang catch
-    # unintentional assignments and suppress all such warnings on MSVC.
-    append_option_if_available("/wd4706" cxx_compile_options)
     # C4848: support for attribute 'msvc::no_unique_address' in C++17 and earlier is a vendor extension
     append_option_if_available("/wd4848" cxx_compile_options)
```
Comment (miscco) [inline cudax/cmake/cudaxBuildCompilerTargets.cmake:21]: We should disable that warning and suppress locally in code

**Review by miscco — COMMENTED** [review]
I like it a lot

File: cmake/CCCLBuildCompilerTargets.cmake (hunk 2 above)
Comment (alliepiper) [reply cmake/CCCLBuildCompilerTargets.cmake:106]: Not sure what you mean -- this line already enables a warning.

File: cudax/cmake/cudaxBuildCompilerTargets.cmake (hunk 4 above)
Comment (alliepiper) [reply cudax/cmake/cudaxBuildCompilerTargets.cmake:21]: Agreed, I see that you've opened #2358 to track this 👍

**Comment by miscco** [comment]
Looks like we are running into https://nvbugspro.nvidia.com/bug/4824645

This should be fixed in a future release

**Comment by alliepiper** [comment]
Closing since all that's left is the c/parallel stuff, which I'll open a new PR for.

**Pull request closed without merging**